proxy/grpcproxy/register.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2017 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcproxy
+
+import (
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/naming"
+	gnaming "google.golang.org/grpc/naming"
+)
+
+type proxyRegister struct {
+	resolverPrefix string
+	c              *clientv3.Client
+	advertiseAddr  string
+	gr             *naming.GRPCResolver
+	leaseID        clientv3.LeaseID
+}
+
+var resolverRefreshSeconds int64 = 60
+
+// Register registers itself as a grpc-proxy server
+// by writing its address prefixed by a given key.
+func Register(c *clientv3.Client, prefix string, addr string, donec chan struct{}) gnaming.Watcher {
+	pr := &proxyRegister{
+		resolverPrefix: prefix,
+		c:              c,
+		advertiseAddr:  addr,
+		gr:             &naming.GRPCResolver{Client: c},
+		leaseID:        clientv3.NoLease,
+	}
+	watcher, err := pr.gr.Resolve(pr.resolverPrefix)
+	if err != nil {
+		plog.Panicf("failed to resolve %q (%v)", pr.resolverPrefix, err)
+	}
+
+	go func() {
+		pr.register()
+		for {
+			select {
+			case <-donec:
```

Comment:
`<-client.Ctx().Done()`? The client ctx is used a lot in the proxy code already.